programs/pluto/cert_rotation.c

```diff
@@ -0,0 +1,50 @@
+#include <nss.h>
+
+#include "cert_rotation.h"
+#include "connections.h"
+#include "constants.h"
+#include "defs.h"
+#include "keys.h"
+#include "log.h"
+#include "nss_cert_load.h"
+#include "whack.h"
+
+void rotate_cert(const struct fd *whackfd, const struct whack_message *wm)
+{
+	struct connection *conn;
+	struct end *dst;
+	const char *nickname;
+	enum whack_pubkey_type pubkey_type;
+	CERTCertificate *old_nss_cert;
+
+	conn = conn_by_name(wm->name, false);
+	if (conn == NULL)  {
+		whack_comment(whackfd, "No connection named '%s' found",
+			wm->name);
+		return;
+	}
+
+	dst = &conn->spd.this;
+	nickname = cert_nickname(&dst->cert);
+	if (nickname == NULL) {
+		whack_comment(whackfd, "No cert nickname found");
+		return;
+	}
+
+	// temporarily back up cert to release or restore it later
+	old_nss_cert = dst->cert.u.nss_cert;
+
+	pubkey_type = WHACK_PUBKEY_CERTIFICATE_NICKNAME;
+	if (!load_left_cert_and_secret(whackfd, nickname, pubkey_type, dst)) {
```

Comment:
What happens if the connection isn't oriented, or itis "right" that is local.

Reply:
This simply loads latest cert and key from NSS DB and put them in pluto's memory.

This doesn't make any changes on an existing connection. The only assumption made here is that the connection is loaded into pluto.

I may be wrong, but I think as far as rotating cert is concerned, it doesn't matter whether the connection is up/down or what is the right end.

Reply:
> This simply loads latest cert and key from NSS DB and put them in pluto's memory.
> 
> This doesn't make any changes on an existing connection. The only assumption made here is that the connection is loaded into pluto.

It changes the existing connection's cert, it also potentially changes the .id, and possibly needs to update .ca (for instance when the original .ca was extracted from the cert, current code doesn't handle this).






> 
> I may be wrong, but I think as far as rotating cert is concerned, it doesn't matter whether the connection is up/down or what is the right end.

At any time a connection is either oriented (left/right somehow map onto that/this) or unoriented (this/that) are arbitrary.  The patch assumes this==left and only end's cert needs an update.  There's nothing to say that .that is the end needing an update.